neutron/convert.cpp

```diff
@@ -97,24 +101,26 @@ auto tofToEnergyConversionFactor(const Dataset &d) {
 
 Dataset tofToEnergy(Dataset &&d) {
   // 1. Compute conversion factor
-  const auto conversionFactor = tofToEnergyConversionFactor(d);
+  const auto conversionFactor =
+      astype(tofToEnergyConversionFactor(d), d.coords()[Dim::Tof].dtype());
```

Comment:
I think you need to keep the conversion factor with its actual `dtype`. For event data there may be multiple things that are converted, and some may be `double` while others are `float`. I think the correct way to do this is as you did above in `convert_with_factor`.